sale_product_set_variant/wizard/product_set_add.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2017-2018 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, fields, api, _
+from odoo.exceptions import UserError, ValidationError
+from odoo.addons import decimal_precision as dp
+
+
+class ProductSetAddLine(models.TransientModel):
+    _name = 'product.set.add.line'
+    _description = 'Product add line'
+    _order = 'sequence'
+
+    wiz_id = fields.Many2one(
+        'product.set.add',
+    )
+    source_line_id = fields.Many2one(
+        'product.set.line',
+    )
+    product_variant_ids = fields.Many2many(
+        'product.product',
+        domain="""[
+            '&',
+            ('sale_ok', '=', True),
+            ('product_tmpl_id', '=', product_template_id),
+        ]""",
+        string='Products',
+    )
+    quantity = fields.Float(
+        string='Quantity',
+        digits=dp.get_precision('Product Unit of Measure'),
+    )
+    product_set_id = fields.Many2one(
+        'product.set',
+        string='Set',
+        related='source_line_id.product_set_id',
+        readonly=True,
+    )
+    product_template_id = fields.Many2one(
+        'product.template',
+        string='Product Template',
+        required=True,
+    )
+    sequence = fields.Integer(
+        string='Sequence',
+        required=True,
+        default=0,
+    )
+
+    @api.onchange('product_template_id')
+    def _onchange_product_template_id(self):
+        for record in self:
+            variants = record.product_template_id.product_variant_ids
+            if len(variants) == 1:
+                record.product_variant_ids = [(6, 0, variants.ids)]
+            else:
+                record.product_variant_ids = [(5, 0, 0)]
+
+
+class ProductSetAdd(models.TransientModel):
+    _inherit = 'product.set.add'
+
+    product_set_id = fields.Many2one(
+        'product.set',
+        'Product set',
+        required=True,
+    )
+    set_line_ids = fields.One2many(
+        'product.set.add.line',
+        'wiz_id',
+        string='Products',
+    )
+
+    @api.onchange('product_set_id')
+    def _onchange_product_set_id(self):
+        vals = []
+        if self.product_set_id:
+            for line in self.product_set_id.set_line_ids:
+                if not line.product_template_id:
+                    raise ValidationError(_(
+                        "No selected product_template_id in Product Set"
```

Comment:
IMO it should be more relevant to user.
Should be like: "Please select Product Template in Product Set" or "Product Template is not selected in Product Set"